lib/babe/verify.go

```diff
@@ -327,62 +327,58 @@ func (b *verifier) verifyAuthorshipRight(header *types.Header) error {
 		return ErrBadSignature
 	}
 
-	// check if the producer has equivocated, ie. have they produced a conflicting block?
-	// hashes is hashes of all blocks with same block number as header.Number
-	hashes := b.blockState.GetAllBlocksAtDepth(header.ParentHash)
+	equivocated, err := b.verifyBlockEquivocation(header)
+	if err != nil {
+		return fmt.Errorf("could not verify block equivocation: %w", err)
+	}
+	if equivocated {
+		return fmt.Errorf("%w for block header %s", ErrProducerEquivocated, header.Hash())
+	}
 
-	for _, currentHash := range hashes {
-		currentHeader, err := b.blockState.GetHeader(currentHash)
-		if err != nil {
-			return fmt.Errorf("failed get header %s", err)
-		}
+	return nil
+}
 
-		currentBlockProducerIndex, err := getAuthorityIndex(currentHeader)
-		if err != nil {
-			return fmt.Errorf("failed to get authority index %s", err)
-		}
+// verifyBlockEquivocation checks if the given block's author has occupied the corresponding slot more than once.
+// It returns true if block was equivocated.
+func (b *verifier) verifyBlockEquivocation(header *types.Header) (bool, error) {
+	author, err := getAuthorityIndex(header)
+	if err != nil {
+		return false, fmt.Errorf("failed to get authority index: %w", err)
+	}
 
-		if len(currentHeader.Digest.Types) == 0 {
-			return fmt.Errorf("current header missing digest")
+	currentHash := header.Hash()
+	slot, err := types.GetSlotFromHeader(header)
+	if err != nil {
+		return false, fmt.Errorf("failed to get slot from header of block %s: %w", currentHash, err)
+	}
+
+	blockHashesInSlot, err := b.blockState.GetBlockHashesBySlot(slot)
+	if err != nil {
+		return false, fmt.Errorf("failed to get blocks produced in slot: %w", err)
+	}
+
+	for _, blockHashInSlot := range blockHashesInSlot {
+		if blockHashInSlot.Equal(currentHash) {
+			continue
 		}
 
-		currentPreDigestItemValue, err := currentHeader.Digest.Types[0].Value()
+		existingHeader, err := b.blockState.GetHeader(blockHashInSlot)
 		if err != nil {
-			return fmt.Errorf("getting current header first digest type value: %w", err)
-		}
-		currentPreDigest, ok := currentPreDigestItemValue.(types.PreRuntimeDigest)
-		if !ok {
-			return fmt.Errorf("%w: got %T", types.ErrNoFirstPreDigest, currentPreDigestItemValue)
+			return false, fmt.Errorf("failed to get header for block %s: %w", blockHashInSlot, err)
```

Comment:
See https://chainsafeio.slack.com/archives/CEATMPULS/p1666965925368239

```suggestion
			return false, fmt.Errorf("failed to get header: %w", err)
```